### ISSUE TITLE
`rustup target remove` does not clean up properly

### ISSUE THREAD

**Issue by badboy** [rust-lang/rustup#596]
After removing a target, the directory sticks around

``` shell
$ cd ~/.multirust
$ rustup target add mips-unknown-linux-musl                  
info: downloading component 'rust-std' for 'mips-unknown-linux-musl'
 16.7 MiB /  16.7 MiB (100 %)   3.0 MiB/s ETA:   0 s                
info: installing component 'rust-std' for 'mips-unknown-linux-musl'
$ ls ./toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/mips-unknown-linux-musl/lib 
liballoc-e8edd0fd.rlib  ...
$ rustup target remove mips-unknown-linux-musl
$ ls ./toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/mips-unknown-linux-musl/lib
$
```

`mips-unknown-linux-musl/lib` is still there, though it is empty.

**Comment by brson** [comment]
This is a bit tricky to solve. It's kinda fundamental to how the package installers work. The individual installers are designed for installing globally. So there are directories they can't just delete (like `/usr/lib`) and the installer doesn't track any difference between directories that rust 'owns' and those it doesn't.

I'd say the easiest way to fix this is to add a flag to `rustup_dist::component::components::Component::uninstall` called `remove_empty_dirs`. During uninstallation it keeps a list of directories it encounters. After uninstall it walks those directories up to the install prefix, deleting them as it goes.

The reason to do it at the end is because uninstall is going to see the same directories over and over and you don't want to keep making the readdir syscalls.

The reason to make it a flag and not just do it by default is because the installer code is *intended* to be independent of the rustup model where the toolchain dirs are owned by rustup.

**Comment by thedrow** [comment]
Shouldn't this be closed since #634 is resolved?

**Comment by kinnison** [comment]
@thedrow Yep, thanks for catching that.
